iProc PCIe Endpoint support

Abhishek-brcm opened the pull request:
iProc PCIe Endpoint IP is present in Broadcom PCIe offload chips. It is primarily used for data transfer between Endpoint device memory and Host memory via iProc PCIe outbound memory.

While zephyr, at some extent, has support for PCIe Root Complex, it does not support PCIe Endpoint currently.

This patch series introduces following public APIs for PCIe endpoint driver (under `pcie_ep.h`):
    - EP configuration space read/write
    - Mapping/Unmapping of Host buffer and PCIe outbound region
    - Raise interrupt to Host
These are minimal base APIs to make any PCIe EP functional.

Based on these APIs, an implementation - iProc PCIe EP driver (`pcie_ep_bcm_iproc.c`) - is added.
Also, a helper API is added (under `pcie_ep_common.c`) to achieve PCIe data transfer with these PCIe EP APIs.
This patch series is tested on Broadcom SoC with iProc PCIe EP IP.

v2:
- Removed config PCIE dependency on x86, and properly placed `Kconfig` and `CMakeLists.txt` include for PCIE EP under `drivers/pcie` instead of directly under `drivers/ `
- Driver renamed as suggested - used subsystem name as prefix
- Fixed comments to be doxygen format compatible
- Other nits resolved
v3:
- Data transfers thoroughly tested between Host and  SoC with iProc PCIe EP IP.
- Instead of using specific property names to obtain outbound memory ranges, used `reg` property as in Linux PCIe EP node.
v4:
- In favor of zephyr choosing C99 data types, replaced zephyr data types, made functions' return type "int" where required.

Abhishek-brcm committed:
pcie: endpoint: Add public APIs for PCIe endpoint driver

Add public APIs for PCIe endpoint driver:
- EP configuration space read/write
- Mapping/Unmapping of Host buffer and PCIe outbound region
- Raise interrupt to Host
These are minimal base APIs to make PCIe EP functional.

Also, add a Kconfig and an empty CMakeLists.txt for drivers to extend.

Signed-off-by: Abhishek Shah <abhishek.shah@broadcom.com>

Abhishek-brcm committed:
pcie: endpoint: Add iProc PCIe EP driver

iProc PCIe EP IP is present in Broadcom PCIe offload chips.
Add iProc PCIe EP driver to provide basic PCIe EP functionality.

Signed-off-by: Abhishek Shah <abhishek.shah@broadcom.com>
Signed-off-by: Shivaraj Shetty <shivaraj.shetty@broadcom.com>

Abhishek-brcm committed:
dts: bindings: Add binding for iProc PCIe EP node

Add device tree binding for the iProc PCIe EP present on
Broadcom Valkyrie/Viper SoC series.

Signed-off-by: Abhishek Shah <abhishek.shah@broadcom.com>

Abhishek-brcm committed:
dts: arm: viper: Add DT node for iProc PCIe EP

dts: arm: Add iProc PCIe EP node in Viper DT

Signed-off-by: Abhishek Shah <abhishek.shah@broadcom.com>
Signed-off-by: Shivaraj Shetty <shivaraj.shetty@broadcom.com>

Abhishek-brcm committed:
pcie: endpoint: Introduce API to achieve PCIe data transfer

Introduce common API to achieve data transfer using memcpy
to/from outbound region of PCIe EP.

Signed-off-by: Abhishek Shah <abhishek.shah@broadcom.com>

dcpleung reviewed (changes requested):
In general, it looks good.

Could you also rename the driver file names to prefix with `pcie_ep_` as file names of drivers are usually prefixed with the subsystem.

Since this is introducing a new API, some requirements are listed here @ https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#experimental .

@carlescufi I assume a presentation in the API meeting is needed to introduce this?

tbursztyka commented on include/drivers/pcie/endpoint/pcie_ep.h:
```diff
@@ -0,0 +1,168 @@
+};
+
+struct pcie_ep_driver_api {
+	u32_t (*conf_read)(struct device *dev, unsigned int offset);
```
and align your types to zephyr types: unsigned int -> u32_t 

Apply that everywhere

Abhishek-brcm replied:
ACK, in that case, will change return type to s32_t.

dcpleung replied:
The data can potentially span the whole range of 0 - 2^31. It would be better to have `int` as return value only, and have an argument to return the actual value.

Abhishek-brcm replied:
ACK

Abhishek-brcm replied:
Done.

tbursztyka commented on include/drivers/pcie/endpoint/pcie_ep.h:
```diff
@@ -0,0 +1,157 @@
+};
+
+/*
+ * @brief       Read PCIe EP configuration space
```
just one space after the doxygen keyword

Abhishek-brcm replied:
ACK

tbursztyka commented on include/drivers/pcie/endpoint/pcie_ep.h:
```diff
@@ -0,0 +1,157 @@
+ * @brief       Read PCIe EP configuration space
+ * @details     This API reads EP's own configuration space
+ *
+ * @param       dev - Pointer to the device structure for the driver instance
```
no need for the '-'

Abhishek-brcm replied:
ACK

tbursztyka commented on drivers/pcie/endpoint/pcie_ep_common.c:
```diff
@@ -0,0 +1,103 @@
+	 * Make sure that address can be generated by core, this condition
+	 * would not hit if proper pcie_ob_mem_type is passed by core
+	 */
+#ifndef CONFIG_64BIT
```
Use `if (!IS_ENABLED(CONFIG_64BIT)) {` instead

Abhishek-brcm replied:
ACK

tbursztyka reviewed (dismissed):
use zephyr types everywhere (s32_t, u32_t etc...)

Abhishek-brcm commented on drivers/CMakeLists.txt:
```diff
@@ -50,3 +50,4 @@ add_subdirectory_ifdef(CONFIG_ENTROPY_HAS_DRIVER entropy)
 add_subdirectory_ifdef(CONFIG_SYS_CLOCK_EXISTS timer)
 add_subdirectory_ifdef(CONFIG_NEURAL_NET_ACCEL neural_net)
 add_subdirectory_ifdef(CONFIG_PTP_CLOCK        ptp_clock)
+add_subdirectory_ifdef(CONFIG_PCIE_ENDPOINT    pcie/endpoint)
```
I thought of the same initially, but `pcie/` directory get included by cmake only when `CONFIG_PCIE` is defined:
`add_subdirectory_if_kconfig(pcie)` [ @ `drivers/CMakeLists.txt`].

And `CONFIG_PCIE=y` can be done only for `x86` as of now (and we are `ARM` platform):
`depends on X86` [ @ `drivers/pcie/Kconfig`]

We should remove this `depends on X86` IMHO and let the platforms select `CONFIG_PCIE` if they want to.

stephanosio replied:
> We should remove this `depends on X86` IMHO and let the platforms select `CONFIG_PCIE` if they want to.

Yes, that is probably what we should do.

Abhishek-brcm replied:
Done part of this PR.

Abhishek-brcm commented on drivers/pcie/endpoint/iproc_pcie.c:
```diff
@@ -0,0 +1,319 @@
+
+	key = k_spin_lock(&ctx->ob_map_lock);
+
+	if (mapped_addr >> 32)
+		ctx->highmem_in_use = false;
+	else
+		ctx->lowmem_in_use = false;
```
ACK

Abhishek-brcm commented on drivers/pcie/endpoint/iproc_pcie.c:
```diff
@@ -0,0 +1,319 @@
+		     &cfg->base->paxb_omap[idx].upper);
+
+	/* Mark usage of outbound window */
+	if (idx == PCIE_MAP_HIGHMEM_IDX)
```
ACK

Abhishek-brcm commented on drivers/pcie/endpoint/pcie_ep_common.c:
```diff
@@ -0,0 +1,103 @@
+		return ret;
+
+	/* Check if we achieved data transfer for given size */
+	if (mapped_size == size)
```
ACK

Abhishek-brcm commented on drivers/pcie/endpoint/pcie_ep_common.c:
```diff
@@ -0,0 +1,103 @@
+				       ob_mem_type);
+
+	/* Check if outbound memory mapping succeeded */
+	if (mapped_size < 0)
```
ACK

Abhishek-brcm commented on drivers/pcie/endpoint/pcie_ep_common.c:
```diff
@@ -0,0 +1,103 @@
+	pcie_ep_unmap_addr(dev, mapped_addr);
+
+	/* Check if mapped_copy succeeded */
+	if (ret < 0)
```
ACK

Abhishek-brcm commented on drivers/pcie/endpoint/pcie_ep_common.c:
```diff
@@ -0,0 +1,103 @@
+		return ret;
+
+	/* In second attempt, we must have completed data transfer */
+	if (mapped_size != unmapped_size)
```
ACK

Abhishek-brcm commented on drivers/Kconfig:
```diff
@@ -29,6 +29,8 @@ source "drivers/entropy/Kconfig"
 
 source "drivers/pcie/Kconfig"
 
+source "drivers/pcie/endpoint/Kconfig"
```
> use zephyr types everywhere (s32_t, u32_t etc...)

ACK

Abhishek-brcm commented on drivers/pcie/endpoint/iproc_pcie_regs.h:
```diff
@@ -0,0 +1,418 @@
+};
+#endif
+
+static inline void pcie_write32(u32_t data, u32_t *addr)
+{
+	sys_write32(data, (mem_addr_t)addr);
+}
+
+static inline u32_t pcie_read32(u32_t *addr)
+{
+	return sys_read32((mem_addr_t)addr);
+}
```
We are using `pcie_write32` and `pcie_read32` just to indicate that we are accessing PCIe registers.

Abhishek-brcm committed:
drivers: pcie: remove depends on X86

Remove depends on X86 for config PCIE,
Let the platforms select PCIE if they want to.

Signed-off-by: Abhishek Shah <abhishek.shah@broadcom.com>

Abhishek-brcm commented on dts/arm/broadcom/viper-common.dtsi:
```diff
@@ -27,5 +27,16 @@
+		pcie0: pcie@4e100000 {
+			compatible = "brcm,iproc-pcie-ep";
+			reg = <0x4e100000 0x2100>;
+			map_lowmem_base = <0x50000000>;
+			map_lowmem_size = <0x8000000>;
+			map_highmem_base = <0x4 0x00000000>;
+			/* 128 MB map is used out of 8 GB mapping region */
+			map_highmem_size = <0x8000000>;
```
Zephyr DT does not seem to support `ranges` or `dma-ranges` property currently in the way that it supports `reg` property. I see that for `reg` property, zephyr generated `*_REG_IDX_0_VAL_ADDRESS` and `*_REG_IDX_0_VAL_SIZE` in accordance with `#address-cells` (= 2) and `#size-cells` (= 2) @`devicetree_unfixed.h` under  `Special property macros` section.  
But the same support is not available for `ranges` or `dma-ranges` property. Please correct me if I am missing anything here.

Trying out this suggestion, after few back and forth, I was able to parse `ranges` property from DT via default provided macros. [I noticed `#address-cells` and `#size-cells` cannot be changed from within the node - unlike Linux DT].  However, I just wanted to set value of few variables in driver through DT - even if I get it done through `ranges` property. So, in the end, just stuck with specific properties(`map_lowmem_*`/`map_highmem_*`), with which I can parse as I want.

stephanosio replied:
cc @mbolivar-nordic @galak

mbolivar-nordic replied:
I believe ranges is used on nRF9160 to correctly pick secure/non-secure peripheral addresses depending on the configuration:

https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/arm/nordic/nrf9160ns.dtsi#L80
https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/arm/nordic/nrf9160_common.dtsi

So some semi-standard ranges support appears to exist. Is there something missing in particular that's blocking its use here?

Abhishek-brcm replied:
The `ranges` property that you pointed out here is not being used in driver, right? This `ranges` property hold the base address and child nodes' `reg` property holds offset from the base,  at the end, (base+offset) value gets generated. -- Basically, used for address translation between child and parent node, also safe-guards the range used by child nodes.

This is different than `ranges` we are talking above. 
For example, Linux supports parsing of `ranges` and `dma-ranges` property under pcie node.
pcie node example:
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/tree/arch/arm64/boot/dts/qcom/msm8996.dtsi#n750
Parsing of `ranges`(for OB resources) and `dma-ranges`(for IB resources) is supported via below function:
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/tree/drivers/pci/of.c#n250
This "ranges" supports passing type of resource(IO/MEM/PREFETCH, etc) that can be mapped to given OB window resource and  PCI OB->AXI translation addresses.
However, all this is w.r.t. PCIe RC, we don't have such requirement for PCIe EP.

For PCIe EP, we would ideally want to just parse the multiple OB region addresses, DT node would look like:
```
pcie@4e100000 {
    .... 
    ranges = <0x0 0x50000000 0x0 0x8000000>,
             <0x4 0x00000000 0x0 0x8000000>;
    ....
};
```
and change driver like:
```
       .map_low_base = DT_INST_PROP_BY_IDX(0, ranges, 0), /* should hold 0x50000000  */
       .map_low_size = DT_INST_PROP_BY_IDX(0, ranges, 1), /* should hold 0x8000000 */
       .map_high_base = DT_INST_PROP_BY_IDX(0, ranges,2), /* should hold 0x400000000*/
       .map_high_size = DT_INST_PROP_BY_IDX(0, ranges, 3), /* should hold 0x8000000 */
```
Currently, I am parsing specific properties from DT, and don't see anything wrong with it. At the end, we just parsing what we need from DT for driver.

@stephanosio please see if this is what you were suggesting.

mbolivar-nordic replied:
Ah, OK, if this is a PCIe specific issue instead of just simple address translation I am not surprised if this isn't supported in dtlib/edtlib.

Abhishek-brcm replied:
I am okay with current approach until above described `ranges` support is available. We can take it up as improvement.
@stephanosio please let us know if you are fine with this..

stephanosio replied:
@Abhishek-brcm Ok, it seems we cannot use `ranges` in Zephyr for now (another TODO for devicetree support, maybe?). 

Marking it as resolved

Abhishek-brcm replied:
I have gone ahead and opened an enhancement here: #25382
Please see if it's okay.

Abhishek-brcm commented on drivers/pcie/endpoint/Kconfig.iproc_pcie:
```diff
@@ -0,0 +1,18 @@
+# Copyright 2020 Broadcom
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig IPROC_PCIE_EP
```
ACK, will change driver name, config, etc. to follow general naming convention.

Abhishek-brcm replied:
Done.

Abhishek-brcm committed:
boards: arm: bcm958402m2_a72: Enable PCIe EP support

Enable PCIe EP support on bcm958402m2_a72 board

Signed-off-by: Abhishek Shah <abhishek.shah@broadcom.com>

Abhishek-brcm commented:
Took liberty to resolve comments related to nits (used braces even for single-line if block, added doxygen format comments properly, etc.) to reduce clutter.

Abhishek-brcm commented:
> use zephyr types everywhere (s32_t, u32_t etc...)

Replaced int by s32_t and unsigned int by u32_t

dcpleung commented:
LGTM. Just need to figure out the devicetree part and we are good to go.

galak commented on dts/bindings/pcie/endpoint/brcm,iproc-pcie-ep.yaml:
```diff
@@ -0,0 +1,39 @@
+      required: true
+      description: register space for the PAX registers
+
+    map_lowmem_base:
```
Based on discussion on slack, map_lowmem_base, map_lowmem_size, map_highmem_base, and map_highmem_size should really be encoded in the reg's property.

Abhishek-brcm replied:
@galak I have used `reg` property to specify PCIe outbound memory ranges now. Please have a look.

galak reviewed (dismissed):
Please change dts to encode the map address in reg.

carlescufi commented:
@galak can you take another look?

carlescufi commented:
@andrewboie @sandeepbrcm @stephanosio @tbursztyka @galak reviews welcome

galak commented on dts/bindings/pcie/endpoint/brcm,iproc-pcie-ep.yaml:
```diff
@@ -0,0 +1,19 @@
+
+properties:
+    reg:
+      type: array
```
you don't need `type` here, its inherited from base.yaml

galak commented on dts/bindings/pcie/endpoint/brcm,iproc-pcie-ep.yaml:
```diff
@@ -0,0 +1,19 @@
+      description: register space for the PAX registers
+
+    label:
+      type: string
```
you don't need `type` here, its inherited from base.yaml

galak commented on dts/bindings/pcie/endpoint/brcm,iproc-pcie-ep.yaml:
```diff
@@ -0,0 +1,19 @@
+    label:
+      type: string
+      required: true
+      description: Human readable string describing the device
```
Don't need description either.

galak commented on dts/bindings/pcie/endpoint/brcm,iproc-pcie-ep.yaml:
```diff
@@ -0,0 +1,19 @@
+    reg:
+      type: array
+      required: true
+      description: register space for the PAX registers
```
Maybe add a bit more about what the PAX registers are here.

galak commented:
LGTM.  Would appreciate @tbursztyka, @andrewboie or @carlescufi take a quick look.

carlescufi reviewed (approved):
LGTM

andrewboie commented on drivers/pcie/endpoint/pcie_ep_bcm_iproc.c:
```diff
@@ -0,0 +1,335 @@
+}
+
+static int iproc_pcie_map_addr(struct device *dev, uint64_t pcie_addr,
+			       uint64_t *mapped_addr, uint32_t size,
```
why not use `size_t` for the size?

Abhishek-brcm replied:
I looked into this. `size_t` is 64-bit wide when building zephyr with LP64 and is 32-bit wide with ILP32. We really want this variable to be just 32-bit wide. So, I would like to stick to `uint32_t`.

andrewboie commented on drivers/pcie/endpoint/pcie_ep_bcm_iproc.c:
```diff
@@ -0,0 +1,335 @@
+	return ret;
+}
+
+static void iproc_pcie_unmap_addr(struct device *dev, uint64_t mapped_addr)
```
question about data types: I see a few places where `uint64_t` is used to store memory addresses, like `mapped_addr` here and elsewhere?

Should we be using `uintptr_t`?
Also I'm interested to know whether these addresses are specifically physical and not virtual; if these are virtual addresses perhaps we could just use `void *`?

Abhishek-brcm replied:
> Should we be using `uintptr_t`?

`mapped_addr`  is PCIe outbound memory address and can be 64-bit disregard of `uintptr_t` as we have this memory in 64-bit range as well. For example, when we build zephyr for cortex-M7 , `uintptr_t` is 32-bit (32-bit wide pointer), but when we build for cortex-A72 , `uintptr_t` is 64-bit (64-bit wide pointer). In both the cases, `mapped_addr`  can actually be 64-bit wide, so  I am using `uint64_t` here.

> Also I'm interested to know whether these addresses are specifically physical and not virtual; if these are virtual addresses perhaps we could just use `void *`?

These addresses are issued by core when MMU is enabled, so these are virtual address. They are 1-to-1 physical-to-virtual address mapped. (mapping is done just to mark these addresses as nGnRnE device memory).
For the same reason as above, using `uint64_t` instead of `void *`.

andrewboie replied:
I think I understand, thanks for the extra explanation. Is it correct to say that this mechanism is for a PCI-e peripheral endpoint and these addresses are from the host CPUs perspective?

Other than my suggestion to use size_t for sizes, no objections to this PR.

Abhishek-brcm replied:
> Is it correct to say that this mechanism is for a PCI-e peripheral endpoint and these addresses are from the host CPUs perspective?

Not all addresses are from the host CPUs perspective. One of the jobs of this driver is to map the Host address to EP's local memory mapped PCIe outbound address. I have put details under public API pcie_ep_map_addr's description, where `pcie_addr` is from host CPUs perspective and `mapped_addr`(mapped outbound address) is from local CPU perspective.

> Other than my suggestion to use size_t for sizes, no objections to this PR.

I will look into this and update.

tyrantlizard77 reviewed (commented):
have a question about data types used for memory addresses.

I'd also prefer if size values were stored using `size_t`.
